pymobiledevice3/remote/core_device_tunnel_service.py

```diff
@@ -0,0 +1,472 @@
+from pymobiledevice3.pair_records import create_pairing_records_cache_folder, generate_host_id
+from pymobiledevice3.remote.remote_pairing_service import RemotePairingService
+from pymobiledevice3.remote.remote_service_discovery import RemoteServiceDiscoveryService
+from pymobiledevice3.remote.remotexpc import XpcInt64Type, XpcUInt64Type
+
+PairingDataComponentType = Enum(Int8ul,
+                                METHOD=0x00,
+                                IDENTIFIER=0x01,
+                                SALT=0x02,
+                                PUBLIC_KEY=0x03,
+                                PROOF=0x04,
+                                ENCRYPTED_DATA=0x05,
+                                STATE=0x06,
+                                ERROR=0x07,
+                                RETRY_DELAY=0x08,
+                                CERTIFICATE=0x09,
+                                SIGNATURE=0x0a,
+                                PERMISSIONS=0x0b,
+                                FRAGMENT_DATA=0x0c,
+                                FRAGMENT_LAST=0x0d,
+                                SESSION_ID=0x0e,
+                                TTL=0x0f,
+                                EXTRA_DATA=0x10,
+                                INFO=0x11,
+                                ACL=0x12,
+                                FLAGS=0x13,
+                                VALIDATION_DATA=0x14,
+                                MFI_AUTH_TOKEN=0x15,
+                                MFI_PRODUCT_TYPE=0x16,
+                                SERIAL_NUMBER=0x17,
+                                MFI_AUTH_TOKEN_UUID=0x18,
+                                APP_FLAGS=0x19,
+                                OWNERSHIP_PROOF=0x1a,
+                                SETUP_CODE_TYPE=0x1b,
+                                PRODUCTION_DATA=0x1c,
+                                APP_INFO=0x1d,
+                                SEPARATOR=0xff)
+
+TLV = Struct(
+    'type' / PairingDataComponentType,
+    'data' / Prefixed(Int8ul, GreedyBytes),
+)
+
+TLVBuf = GreedyRange(TLV)
+
+PairConsentResult = namedtuple('PairConsentResult', 'public_key salt')
+
+CDTunnelPacket = Struct(
+    'magic' / Const(b'CDTunnel'),
+    'body' / Prefixed(Int16ub, GreedyBytes),
+)
+
+logger = logging.getLogger(__name__)
+
+
+class RemotePairingTunnel(QuicConnectionProtocol):
+    def __init__(self, quic: QuicConnection, stream_handler: Optional[QuicStreamHandler] = None):
+        super().__init__(quic, stream_handler)
+        self._ack_waiter: Optional[asyncio.Future[None]] = None
+
+    async def request_tunnel_establish(self) -> Mapping:
+        stream_id = self._quic.get_next_available_stream_id()
+
+        # TODO: understand what this buffer should actually do
+        self._quic.send_datagram_frame(b'x' * 1024)
+
+        self._quic.send_stream_data(stream_id, self._encode_cdtunnel_packet(
+            {'type': 'clientHandshakeRequest', 'mtu': 1420}))
+
+        waiter = self._loop.create_future()
```

Comment:
1. If the result it required, I would put it in a high level object such as queue, otherwise an event
2. I would move it to the constructor

Reply:
done